pyblish/plugin.py

```diff
@@ -462,10 +503,38 @@ def process(plugin, context, instance=None, action=None):
 
     """
 
-    if issubclass(plugin, (ContextPlugin, InstancePlugin)):
-        result = __explicit_process(plugin, context, instance, action)
-    else:
-        result = __implicit_process(plugin, context, instance, action)
+    if not context.get('profiling'):
+        context.data['profiling'] = {}
+
+    profiler = ProfilingTimer()
+    with profiler:
+        if issubclass(plugin, (ContextPlugin, InstancePlugin)):
+            result = __explicit_process(plugin, context, instance, action)
+        else:
+            result = __implicit_process(plugin, context, instance, action)
+
+    profiling_data = {
+        "runtime": profiler.runtime,
+    }
+
+    # collected using psutil module, None if it is not installed
+    try:
+        profiling_data['initial_memory_sage'] = profiler.initial_mem
+        profiling_data['final_memory_usage'] = profiler.final_mem
+    except AttributeError:
+        # no memory information collected
+        pass
+
+    context.data['profiling'][plugin] = profiling_data
+    profiling_data["snapshot"] = {
+        "timestamp": int(time.time()),
+        "context": json.dumps(context, default=lambda o: '<not serializable>')
```

Comment:
Not sure about "snapshot".. If the goal is capturing more data, I think we should instead favour a mechanism with which to let the plug-in author provide his own data, as mentioned in https://github.com/pyblish/pyblish-base/issues/348#issuecomment-527023863

Reply:
I was thinking about snapshot as "map" if you will, that with all those profiling data you could later on visualize whole pyblish process with all its timings and what was changed in context each step. But yes, adding custom data from every plugin should be implemented also.